Add affinity and tolerations APIManager configurability for DeploymentConfigs

miguelsorianod opened the pull request:
This PR intends to add configurability of affinity and tolerations in APIManager for DeploymentConfigs (database and non-database ones)

The implementation has the following remarkable points:
* Reconciliation of 'Affinity' and 'Tolerations' has been implemented for all DeploymentConfigs. This means the reconciliation is executed even if the DeploymentConfig does not have configurability at APIManager level and even if the DeploymentConfig is a database one. It still should work without issues as the default values for those are nil and no changes should be detected. The only new difference in behaviour for database DCs and DCs where APIManager configurability is not provided is that if the user manually changes the values for them directly in the DCs the changes will be reverted by the operator to nil for them.
* Affinity and Tolerations are deep objects (several levels and multiple fields) and contain arrays (where elements order matters). Reconciliation has been implemented with a DeepEqual comparison and seems to work ok. Changing the order of the specified array elements in APIManager is detected as a change even if the elements are the same, so a reconciliation happens in that case too. This behaviour is no different than changing array elements in a DeploymentConfig in OpenShift.
* Operator-only functionality

eguzki commented:
Looks go. 
Go ahead and implement for all 3scale components. 

For database DC as well.

eguzki commented on pkg/reconcilers/deploymentconfig.go:
```diff
@@ -43,6 +44,12 @@ func GenericDeploymentConfigMutator(existingObj, desiredObj common.KubernetesObj
 	tmpUpdate = DeploymentConfigContainerResourcesReconciler(desired, existing)
 	update = update || tmpUpdate
 
+	tmpUpdate = DeploymentConfigAffinityReconciler(desired, existing)
```
Why not Install Only?

miguelsorianod replied:
It could be install only too I guess.
The parameters are inherently configurable so I decided to implement it configurable.

miguelsorianod commented:
I've added fields and implementations for both non-db and db DCs.

As originally commented in issue `THREESCALE-5053` there are some elements that do not fit in any specific preexisting section:

system-memcached. I've added two fields in the main 'System' section:
system.memcachedAffinity
system.memcachedTolerations

system-sphinx. I've added a new section called 'sphinx' in 'system'
system.sphinx.affinity
system.sphinx.tolerations

zync-database. I've added two fields in the main 'Zync' section:
zync.databaseAffinity
zync.databaseTolerations

The reasons why sphinx has its own section and the others don't are:
* I considered the other ones 'databases' and not sphinx. To be honest I think it could be argued whether sphinx is a database or not or whether potentially could be external or not. It's in sort of a middleground between application and a database.
* I'm thinking on memcached and zync-database potentially being external (zync-database has been requested by some people in the past) so I didn't want to create a new section for fields that might only apply in an 'embedded-only' context. If the PR https://github.com/3scale/3scale-operator/pull/383 related to increased database granularity was performed I'd just have created sections for them too and add them inside an 'embedded' subsection but it's not the case as the changes there have been at least temporarily delayed so I decided to add them directly in the main section and deprecate them in the future if we end up adding granularity for embedded/external DBs.
* zync-database and system-memcached already have fields in their corresponding main-level component (postgresqlimage field and memcachedImage field)

I also have some doubts regarding the naming of memcached and zync postgresql. If you look at the Jira issue `THREESCALE-5053` I already expressed my concern:

memcached:  maybe a more generic term name than memcached that is not limited to memcached implementation in case it changes in the future (for example being in redis instead of memcached)? It's true though that we already have memcached-named fields (memcachedImage)
sphinx: maybe a more generic term name than sphinx that is not limited to sphinx implementation in case it changes in the future?
zync database: maybe call it databaseSpec.postgresSQL? maybe postgreSQL? will it potentially change or have possible different dbs? We already have a postgreSQLImage field but might be a good idea to just name it databaseXXX (although we would end up having different namings between the image and the affinity configs then).

Finally, keep in mind that depending on what changes are performed it might additional code migrations (due to possible deprecations of fields, changes in current schema, ...)

Please let's review the current implementation before continuing.

Thank you.

eguzki commented:
For unittest, you can "initialize" your apimanager objects calling `SetDefaults`. Then the context will be closer to production code running.

eguzki commented on deploy/crds/apps.3scale.net_apimanagers_crd.yaml:
```diff
@@ -41,19 +41,1353 @@ spec:
                   type: boolean
                 productionSpec:
                   properties:
+                    affinity:
```
Definitely this CRD file is not human readable anymore. It is already too big. 

We have to trust on tooling

miguelsorianod commented on pkg/3scale/amp/operator/system_mysql_reconciler.go:
```diff
@@ -25,7 +25,7 @@ func (r *SystemMySQLReconciler) Reconcile() (reconcile.Result, error) {
 	}
 
 	// DC
-	err = r.ReconcileDeploymentConfig(systemMySQL.DeploymentConfig(), reconcilers.CreateOnlyMutator)
```
Using reconcilers.CreateOnlyMutator seemed originally wrong, as  SystemPostgreSQL Reconciler was using `DeploymentConfigResourcesMutator` instead of `reconcilers.CreateOnlyMutator`.

In new reconciler used in this case I've made it reconcile resources too

eguzki replied:
:+1:

miguelsorianod commented:
New changes made:
* Updated mutators used for DCs that weren't using the Generic Mutator
* Added unit tests
* Updated documentation

Ready for review @eguzki

eguzki commented:
Overall LGTM.

I miss some documentation example. Updating API reference is not enough.

Something like, let's configure: 
* node affinity A for apicast, backend listener and backend redis
* node affinity B for the remaining components

I guess this is going to be useful eventually.

miguelsorianod commented:
> Overall LGTM.
> 
> I miss some documentation example. Updating API reference is not enough.
> 
> Something like, let's configure:
> 
>     node affinity A for apicast, backend listener and backend redis
>     node affinity B for the remaining components
> 
> I guess this is going to be useful eventually.

Being advanced use cases I considered there was not much point on having specific examples for it but I've added an example anyway :+1:
